File: frontend/src/lib/passwordTransport.ts
/**
 * 비밀번호 전송 봉인(2026-09-07) -- 서버 src/dms/api/password_transport.py 의 거울.
 *
 * 포탈은 비밀번호를 **평문으로 보내지 않는다**. 서버 공개키(GET /api/auth/transport-key)
 * 와 브라우저 임시 P-256 키로 ECDH → HKDF-SHA256 → AES-256-GCM 봉인을 만들어
 * `password_enc` 로 보낸다. TLS 위에 한 겹 더인 이유: TLS 는 ingress 에서 끝나고
 * 그 뒤 클러스터 내부 hop·TLS 검사 프록시·인증서 경고를 무시한 접속에서는 본문의
 * 평문이 보인다(서버 모듈 docstring 의 위협 모델). AAD 에 용도·사용자명을 묶어
 * 다른 엔드포인트·다른 계정으로의 재사용을 막는다.
 *
 * 상수(salt/info/AAD 접두)와 절차는 서버와 **바이트 단위로** 같아야 한다 -- 한쪽만
 * 바꾸면 전 로그인이 password_encryption_invalid 다.
 *
 * WebCrypto(crypto.subtle) 는 보안 컨텍스트(HTTPS 또는 localhost)에서만 있다. 없으면
 * 평문으로 떨어지지 않고 password_encryption_unavailable 로 멈춘다 -- DMS 는 HTTPS
 * 전용(Secure 쿠키)이라 정상 접속에서는 일어나지 않는다.
 */
import { ApiError, apiGet, apiSend, reasonText } from "./api";

export const TRANSPORT_VERSION = 1;
export const HKDF_SALT = "dms-password-transport";
export const SESSION_KEY_INFO = "dms-password-transport-v1/aes-256-gcm";
export const AAD_PREFIX = "dms-password-transport-v1";

/** 비밀번호를 받는 엔드포인트마다 하나 -- 서버 PURPOSES 와 같은 값. */
export type Purpose = "login" | "signup" | "password_reset" | "admin_create";

export interface TransportKey { version: number; kid: string; public_key: string }
export interface EncryptedPassword {
  version: number; kid: string; epk: string; iv: string; ct: string;
}

const utf8 = (s: string) => new TextEncoder().encode(s);

export function b64encode(bytes: ArrayBuffer | Uint8Array): string {
  const arr = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let bin = "";
  for (const b of arr) bin += String.fromCharCode(b);
  return btoa(bin);
}

// 반환 타입을 ArrayBuffer 기반으로 고정한다 -- TS 5.7+ 의 Uint8Array<ArrayBufferLike>
// 는 WebCrypto 의 BufferSource 에 대입되지 않는다(SharedArrayBuffer 가능성).
export function b64decode(s: string): Uint8Array<ArrayBuffer> {
  const bin = atob(s);
  const out = new Uint8Array(new ArrayBuffer(bin.length));
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function aadFor(purpose: Purpose, username: string): string {
  return `${AAD_PREFIX}|${purpose}|${username}`;
}

export function transportAvailable(): boolean {
  return typeof crypto !== "undefined" && crypto !== null && !!crypto.subtle;
}

/** 순수 봉인 -- fetch 없음. 테스트가 임의 키로 검증하고 e2e 헬퍼가 Node 에서 재사용한다. */
export async function encryptWithKey(
  key: TransportKey, password: string, purpose: Purpose, username: string,
  subtle: SubtleCrypto = crypto.subtle,
): Promise<EncryptedPassword> {
  const serverKey = await subtle.importKey(
    "raw", b64decode(key.public_key), { name: "ECDH", namedCurve: "P-256" }, false, []);
  const ephemeral = await subtle.generateKey(
    { name: "ECDH", namedCurve: "P-256" }, true, ["deriveKey", "deriveBits"]);
  const shared = await subtle.deriveBits(
    { name: "ECDH", public: serverKey }, ephemeral.privateKey, 256);
  const hkdfKey = await subtle.importKey("raw", shared, "HKDF", false, ["deriveKey"]);
  const aes = await subtle.deriveKey(
    { name: "HKDF", hash: "SHA-256", salt: utf8(HKDF_SALT), info: utf8(SESSION_KEY_INFO) },
    hkdfKey, { name: "AES-GCM", length: 256 }, false, ["encrypt"]);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const ct = await subtle.encrypt(
    { name: "AES-GCM", iv, additionalData: utf8(aadFor(purpose, username)) },
    aes, utf8(password));
  const epk = await subtle.exportKey("raw", ephemeral.publicKey);
  return { version: TRANSPORT_VERSION, kid: key.kid,
           epk: b64encode(epk), iv: b64encode(iv), ct: b64encode(ct) };
}

// 서버 키 캐시. 시크릿 회전으로 키가 바뀌면 서버가 password_encryption_key_mismatch
// 를 내고, 아래 postWithSealedPassword 가 캐시를 비운 뒤 한 번 재시도한다.
let cachedKey: Promise<TransportKey> | null = null;

export function fetchTransportKey(): Promise<TransportKey> {
  if (cachedKey === null) {
    cachedKey = apiGet<TransportKey>("/api/auth/transport-key").catch((e) => {
      cachedKey = null;   // 실패한 약속을 캐시하면 영원히 실패한다
      throw e;
    });
  }
  return cachedKey;
}

export function forgetTransportKey(): void {
  cachedKey = null;
}

/**
 * 본문의 `password` 를 `password_enc` 로 바꿔 POST 한다 -- 비밀번호를 보내는 훅
 * 넷(login/signup/password-reset/admin accounts)이 전부 이 함수를 쓴다. 새 훅이
 * 생기면 여기로 -- apiSend 에 password 를 직접 실으면 그 경로만 평문이다.
 */
export async function postWithSealedPassword<T>(
  path: string, purpose: Purpose,
  body: { username: string; password: string } & Record<string, unknown>,
): Promise<T> {
  if (!transportAvailable()) {
    throw new ApiError(0, "password_encryption_unavailable",
                       reasonText("password_encryption_unavailable"));
  }
  const { password, ...rest } = body;
  const attempt = async (): Promise<T> => {
    const key = await fetchTransportKey();
    const password_enc = await encryptWithKey(key, password, purpose, body.username);
    return apiSend<T>("POST", path, { ...rest, password_enc });
  };
  try {
    return await attempt();
  } catch (e) {
    if (e instanceof ApiError && e.code === "password_encryption_key_mismatch") {
      forgetTransportKey();
      return attempt();
    }
    throw e;
  }
}

File: src/dms/api/password_transport.py
"""브라우저 → API 비밀번호 **전송 봉인**(2026-09-07, 사용자 결정 "사용자 비밀번호 암호화").

왜 TLS 로 끝나지 않는가: TLS 는 사용자 PC 와 ingress-nginx 사이만 지킨다. ingress 가
TLS 를 벗긴 뒤 dms-api 까지의 클러스터 내부 hop 은 평문이고, TLS 검사 프록시(사내
방화벽)·인증서 경고를 무시하고 들어온 접속·프록시 본문 로그 어디서든 요청 본문의
평문 비밀번호가 보인다. 그래서 비밀번호 **필드만** 브라우저에서 한 번 더 봉인해
dms-api 프로세스 안에서만 열리게 한다. 저장은 여기서 열린 평문을 그대로
accounts.py 가 scrypt 로 해시한다 -- 평문은 요청 처리 스택 프레임 밖으로 나가지 않는다.

구성(ECIES 류, 전부 표준 프리미티브 -- 자체 설계 암호 없음):
  서버 정적 P-256 키 + 브라우저 임시 P-256 키 → ECDH → HKDF-SHA256 → AES-256-GCM.
  서버 키는 DMS_SESSION_SECRET 에서 HKDF 로 **결정적으로 유도**한다: 레플리카가
  여럿이어도 같은 키, DB 테이블·파일 없음, 시크릿 회전이 곧 키 회전(세션 시크릿이
  새면 세션 위조가 이미 가능하므로 이 키를 같은 비밀에 묶어도 위협이 늘지 않는다).
  AAD 에 용도·사용자명을 묶는다 -- 로그인용 봉인을 계정 생성에, alice 용을 bob 에
  재사용할 수 없다.

위협 모델(정직하게): 수동 도청·TLS 종단 이후 평문·프록시/로그 노출을 막는다.
페이지 JS 자체를 바꿔치기할 수 있는 **완전한 능동 MITM 은 막지 못한다** -- 그건
인증서 신뢰(사용자 CA 설치 또는 공인 CA)의 몫이다. 재전송(replay)도 막지 않는다:
봉인을 가로챌 위치의 공격자는 응답의 세션 쿠키도 가로채므로 replay 방어가 더해
주는 것이 없고, 시각 검사는 PC 시계가 틀린 사용자의 로그인을 깨뜨린다.

와이어 형식(요청 본문 password_enc):
  {"version": 1, "kid": <서버 공개키 지문 16hex>, "epk": b64(65B 비압축점),
   "iv": b64(12B), "ct": b64(암호문+GCM 태그)}
공개키는 GET /api/auth/transport-key 로 받는다(무인증 -- 공개키는 비밀이 아니다).
kid 가 서버와 다르면 password_encryption_key_mismatch: 시크릿 회전 직후 브라우저가
옛 키를 캐시한 경우이며 프런트는 키를 다시 받아 한 번 재시도한다.

프런트 frontend/src/lib/passwordTransport.ts 가 이 모듈의 거울이다 -- 상수(salt/
info/AAD 접두)와 절차가 **바이트 단위로** 같아야 한다. 한쪽만 바꾸면 전 로그인이
password_encryption_invalid 로 죽는다(test_password_transport 가 양쪽 상수를 대조).
"""
import base64
import hashlib
import os

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

VERSION = 1
# 비밀번호를 받는 엔드포인트마다 하나 -- AAD 로 묶여 다른 용도로 재사용이 안 된다.
PURPOSES = ("login", "signup", "password_reset", "admin_create")

_CURVE = ec.SECP256R1()
# P-256 군의 위수(n). 유도한 32 바이트를 [1, n-1] 로 접어 유효한 개인키 스칼라로 만든다.
_ORDER = 0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551
# 아래 세 상수는 프런트 passwordTransport.ts 와 동일해야 한다.
HKDF_SALT = b"dms-password-transport"
SERVER_KEY_INFO = b"dms-password-transport-v1/server-key"
SESSION_KEY_INFO = b"dms-password-transport-v1/aes-256-gcm"
AAD_PREFIX = "dms-password-transport-v1"
_IV_BYTES = 12
_POINT_BYTES = 65


class PasswordTransportError(Exception):
    """복호 실패. reason_code 는 그대로 HTTP 422 detail 이 된다(사유 코드 계약)."""

    def __init__(self, *, reason_code: str):
        self.reason_code = reason_code
        super().__init__(reason_code)


def _hkdf(info: bytes) -> HKDF:
    return HKDF(algorithm=hashes.SHA256(), length=32, salt=HKDF_SALT, info=info)


def _b64d(value) -> bytes:
    if not isinstance(value, str):
        raise ValueError("not a string")
    return base64.b64decode(value, validate=True)


def _b64e(raw: bytes) -> str:
    return base64.b64encode(raw).decode("ascii")


def aad_for(purpose: str, username: str) -> bytes:
    return f"{AAD_PREFIX}|{purpose}|{username}".encode("utf-8")


def key_id(public_key_raw: bytes) -> str:
    return hashlib.sha256(public_key_raw).hexdigest()[:16]


def derive_private_key(secret: str) -> ec.EllipticCurvePrivateKey:
    """세션 시크릿 → P-256 개인키(결정적). 같은 시크릿이면 어느 프로세스든 같은 키."""
    seed = _hkdf(SERVER_KEY_INFO).derive(secret.encode("utf-8"))
    scalar = int.from_bytes(seed, "big") % (_ORDER - 1) + 1
    return ec.derive_private_key(scalar, _CURVE)


def _public_raw(key: ec.EllipticCurvePublicKey) -> bytes:
    return key.public_bytes(serialization.Encoding.X962,
                            serialization.PublicFormat.UncompressedPoint)


class PasswordTransport:
    def __init__(self, secret: str):
        self._private = derive_private_key(secret)
        self.public_key_raw = _public_raw(self._private.public_key())
        self.kid = key_id(self.public_key_raw)

    def public_info(self) -> dict:
        """GET /api/auth/transport-key 응답 -- 프런트 TransportKey 와 같은 모양."""
        return {"version": VERSION, "kid": self.kid,
                "public_key": _b64e(self.public_key_raw)}

    def decrypt(self, payload, *, purpose: str, username: str) -> str:
        """봉인을 열어 평문 비밀번호를 돌려준다. 실패는 전부 PasswordTransportError.

        kid 불일치만 따로 구분한다(프런트가 키를 다시 받아 재시도할 유일한 경우).
        나머지 -- 형식·점·태그·인코딩 -- 는 한 사유로 뭉갠다: 세분하면 오류 메시지가
        복호 오라클이 된다."""
        if not isinstance(payload, dict) or payload.get("version") != VERSION:
            raise PasswordTransportError(reason_code="password_encryption_invalid")
        if payload.get("kid") != self.kid:
            raise PasswordTransportError(reason_code="password_encryption_key_mismatch")
        try:
            epk = _b64d(payload.get("epk"))
            iv = _b64d(payload.get("iv"))
            ct = _b64d(payload.get("ct"))
            if len(iv) != _IV_BYTES or len(epk) != _POINT_BYTES:
                raise ValueError("bad lengths")
            peer = ec.EllipticCurvePublicKey.from_encoded_point(_CURVE, epk)
            shared = self._private.exchange(ec.ECDH(), peer)
            key = _hkdf(SESSION_KEY_INFO).derive(shared)
            plain = AESGCM(key).decrypt(iv, ct, aad_for(purpose, username))
            return plain.decode("utf-8")
        except Exception:
            # ValueError/TypeError/binascii.Error/InvalidTag/UnicodeDecodeError --
            # 어느 것도 클라이언트에 구분해 알릴 이유가 없다(위 docstring).
            raise PasswordTransportError(reason_code="password_encryption_invalid")


def seal(public_key_raw: bytes, password: str, *, purpose: str, username: str) -> dict:
    """브라우저 쪽 봉인의 파이썬 판 -- 테스트와 운영 스크립트(curl 대신)가 쓴다.
    프런트 encryptWithKey 와 절차가 같다(임시 키 → ECDH → HKDF → AES-GCM)."""
    ephemeral = ec.generate_private_key(_CURVE)
    peer = ec.EllipticCurvePublicKey.from_encoded_point(_CURVE, public_key_raw)
    shared = ephemeral.exchange(ec.ECDH(), peer)
    key = _hkdf(SESSION_KEY_INFO).derive(shared)
    iv = os.urandom(_IV_BYTES)
    ct = AESGCM(key).encrypt(iv, password.encode("utf-8"), aad_for(purpose, username))
    return {"version": VERSION, "kid": key_id(public_key_raw),
            "epk": _b64e(_public_raw(ephemeral.public_key())),
            "iv": _b64e(iv), "ct": _b64e(ct)}


def seal_with_info(info: dict, password: str, *, purpose: str, username: str) -> dict:
    """public_info() 응답(또는 GET /api/auth/transport-key 본문)으로 바로 봉인한다."""
    return seal(base64.b64decode(info["public_key"]), password,
                purpose=purpose, username=username)
